Refactor BoxOps tests to use parameterize

HeungwooLee opened the pull request:
<!-- Before submitting a PR, please make sure to check our contributing guidelines regarding code formatting, tests, and documentation: https://github.com/pytorch/vision/blob/main/CONTRIBUTING.md -->
Closes #4500 

Refactored TestBoxArea, TestBoxIoU and TestGenBoxIoU class to not have duplicated code.

* Tests
  * formatting and typing: got warnings and errors but not from the change of this commit
  * `pytest test -vvv`: passed

HeungwooLee commented:
1st commit was failed with lint failure ``Format files with µfmt``
However, as there wasn't anything mentioned required this on CONTRIBUTING.md, I have go ahead and followed instruction from https://ufmt.omnilib.dev/en/stable/ but got more failures. Need some guidance to address failures.

datumbox commented:
@HeungwooLee Thanks! I will have a closer look next week.

Just not that the failing [linter](https://app.circleci.com/pipelines/github/pytorch/vision/14605/workflows/0ca7ec5e-5431-4c49-b4d7-42a0f2872c86/jobs/1175203) job is related to you PR. Don't worry about `binary_win_wheel_py3.9_cu115` it's not broken due to your PR.

HeungwooLee commented on test/test_ops.py:
```diff
@@ -1107,114 +1107,149 @@ def test_bbox_convert_jit(self):
+        [285.25, 188.75, 1192.0, 851.0],
+        [279.25, 198.0, 1189.0, 849.0],
+    ]
+    __FLOAT_16_EXPECTED_OUTPUT = [605113.875, 600495.1875, 592247.25]
```
I am fine with either approach. If method is preferred approach on this repo, I should follow. Will update in next revision.

HeungwooLee commented on test/test_ops.py:
```diff
@@ -1107,114 +1107,149 @@ def test_bbox_convert_jit(self):
+            pytest.param(__INT_TEST_INPUT, torch.int8, __DEFAULT_TOLERANCE, __INT_EXPECTED_OUTPUT),
+            pytest.param(__INT_TEST_INPUT, torch.int16, __DEFAULT_TOLERANCE, __INT_EXPECTED_OUTPUT),
+            pytest.param(__INT_TEST_INPUT, torch.int32, __DEFAULT_TOLERANCE, __INT_EXPECTED_OUTPUT),
+            pytest.param(__INT_TEST_INPUT, torch.int64, __DEFAULT_TOLERANCE, __INT_EXPECTED_OUTPUT),
```
I am a bit confused on this comment. If what you are suggesting here is to simply replace line 1168 to 1171 to be ```pytest.param(__INT_TEST_INPUT, [torch.int8, torch.int16, ...], __DEFAULT_TOLERANCE, __INT_EXPECTED_OUTPUT)``` then this result in single test run with dtypes array as parameter. This doesn't matches goal here as I then need to run loop in single test run to verify 4 different types, which is should be better parameterised.

If what you are suggesting here is to have dtypes parameterised seperated like ```@pytest.mark.parametrize("dtype", pytest.param([torch.int8, torch.int16, ...])``` then I found this runs as cartesian product of all possible combination of parametrised parameters, which includes incorrect combination as float input with int dtypes. The only way to avoid this is to have define separate method for int and float, which I didn't choose because it will create some level of duplication.

But I do see your concern here with duplicated params. I will see what I can do here.

datumbox replied:
Sorry I should have been clearer. I was implying that you will replace the `dtype` with `dtypes` (plural) and refactor the test to loop for each provided type within the list. This is an optional nit that we can leave to decide at the very end. :)

HeungwooLee commented on test/test_ops.py:
```diff
@@ -1107,114 +1107,149 @@ def test_bbox_convert_jit(self):
+        scripted_fn = torch.jit.script(target_fn)
+        return scripted_fn(box, box) if is_binary_fn else scripted_fn(box)
 
+    def _script_box_operation(self, box: Tensor):
```
1. No problem. I can make them as public method.
2. Sorry about the mess and confusion, the PR was submitted with interim change. You are right about these two. As name says these meant to replace two different ways of api calls as shown below.
  * `_perform_box_operation` to replace
    * `ops.box_area(box)`
    * `ops.box_iou(box, box)`
    * `ops.generalized_box_iou(box, box)`
  * `_script_box_operation ` to replace `torch.jit.script(...)` calls

Even though codes are similar, what I need to call at the end are different and I can't think of merging them into one. I will try reduce boilerplates though.

datumbox replied:
> No problem. I can make them as public method.

You should leave the private. My point was that it's a private method and it wasn't used anywhere so something was wrong.

> Even though codes are similar, what I need to call at the end are different and I can't think of merging them into one. 

One approach would be to remove the specific method and make `_perform_box_operation` accept a boolean flag which indicates whether the method should be scripted internally.

HeungwooLee replied:
> One approach would be to remove the specific method and make _perform_box_operation accept a boolean flag which indicates whether the method should be scripted internally.

I also had that approach in mind but worried if that's will violated [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) base on how code reader sees it. calling box operation and call it through `jit.scipt()` can look different base on perspective.

I will try come up with good approach to make them look nit and nicer.

HeungwooLee commented on test/test_ops.py:
```diff
@@ -1107,114 +1107,149 @@ def test_bbox_convert_jit(self):
-        scripted_iou = scripted_fn(box_tensor, box_tensor)
-        torch.testing.assert_close(scripted_iou, expected, rtol=0.0, atol=TOLERANCE)
+        expected = self._perform_box_operation(box_tensor)
+        scripted_area = self._perform_box_operation(box_tensor)
```
You are correct! Will correct this in next revision.

HeungwooLee commented on test/test_ops.py:
```diff
@@ -1107,114 +1107,149 @@ def test_bbox_convert_jit(self):
-            torch.testing.assert_close(out, expected, rtol=0.0, check_dtype=False, atol=tolerance)
+class TestBoxArea(BoxTestBase):
+    __DEFAULT_TOLERANCE = 1e-4
+    __TOLERANCE_0_DOT_05 = 0.05
```
Sure

HeungwooLee commented:
Just updated with 3rd commit which should address all comments above.

* formatting: `ufmt format test/test_ops.py`
* test: `pytest test/test_ops.py -vvv`: passed
* type check: `mypy --config-file mypy.ini`: this failed with error not related to change `torchvision/models/quantization/utils.py:50: error: Module has no attribute "fuse_modules_qat"; maybe "fuse_modules"?`

HeungwooLee commented:
Addressed typing errors and updated PR with new commit.

* formatting: `ufmt format test/test_ops.p` then manually reverted 2 changes with `**` operator
* typing: `mypy --config-file mypy.ini test/test_ops.p`
  * 13 [misc] errors complaining about methods used inside `pytest.param(...)` doesn't have at least self argument or being static class. As other examples in file were ignoring this error, I assume this is somewhat acceptable. Let me know if this is still an issue.
* test: `pytest test/test_ops.py -vvv` (passed)
* automated checks: I believe two failures are not related to my change.

datumbox commented on test/test_ops.py:
```diff
@@ -1107,114 +1107,149 @@ def test_bbox_convert_jit(self):
+    def _target_fn(self) -> Tuple[bool, Callable]:
+        pass
 
+    def __perform_box_operation(self, box: Tensor, run_as_script: bool = False) -> Tensor:
```
To avoid name mangling, I would recommend changing these to `_perform_box_operation`. Same for the rest of the methods that start with double underscore as we don't use this idiom in the codebase.

datumbox reviewed (commented):
LGTM, only one final comment and we should be good:

datumbox commented:
@HeungwooLee Thanks for the awesome work, LGTM.
